## Part 1: Preprocessing

In [13]:
# Import our dependencies
import tensorflow as tf
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Input, Dense
from sklearn.preprocessing import OneHotEncoder


#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [14]:
# Determine the number of unique values in each column.
attrition_df.nunique().sort_values(ascending=False)


HourlyRate                  71
Age                         43
TotalWorkingYears           40
YearsAtCompany              37
DistanceFromHome            29
YearsInCurrentRole          19
YearsWithCurrManager        18
YearsSinceLastPromotion     16
PercentSalaryHike           15
NumCompaniesWorked          10
JobRole                      9
TrainingTimesLastYear        7
EducationField               6
JobLevel                     5
Education                    5
EnvironmentSatisfaction      4
JobInvolvement               4
JobSatisfaction              4
RelationshipSatisfaction     4
StockOptionLevel             4
WorkLifeBalance              4
Department                   3
BusinessTravel               3
MaritalStatus                3
OverTime                     2
Attrition                    2
PerformanceRating            2
dtype: int64

In [15]:
# Create y_df with the Attrition and Department columns
y_df = attrition_df[['Attrition', 'Department']]
y_df.head()

Attrition              Department
0       Yes                   Sales
1        No  Research & Development
2       Yes  Research & Development
3        No  Research & Development
4        No  Research & Development

In [16]:
# Create a list of at least 10 column names to use as X data
X_df = attrition_df[['HourlyRate', 'Age', 'PerformanceRating',
                               'YearsAtCompany','DistanceFromHome',
                                'EnvironmentSatisfaction',
                               'YearsWithCurrManager','YearsSinceLastPromotion',
                               'PercentSalaryHike','MaritalStatus' ]]
X_df.head()


# Create X_df using your selected columns


# Show the data types for X_df
print(X_df.dtypes)



HourlyRate                  int64
Age                         int64
PerformanceRating           int64
YearsAtCompany              int64
DistanceFromHome            int64
EnvironmentSatisfaction     int64
YearsWithCurrManager        int64
YearsSinceLastPromotion     int64
PercentSalaryHike           int64
MaritalStatus              object
dtype: object


In [17]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_df, y_df, random_state=1)


In [18]:
# Convert your X data to numeric data types however you see fit
# Add new code cells as necessary
ohe = OneHotEncoder(handle_unknown='ignore', sparse=False).set_output(transform="pandas")
ohetransform = ohe.fit_transform(X_train[['MaritalStatus','PerformanceRating','EnvironmentSatisfaction']])
ohetransform.head()
X_train = pd.concat([X_train, ohetransform], axis=1).drop(columns=['MaritalStatus','PerformanceRating','EnvironmentSatisfaction'])
X_train.head()

ohetransform = ohe.fit_transform(X_test[['MaritalStatus','PerformanceRating','EnvironmentSatisfaction']])
ohetransform.head()
X_test = pd.concat([X_test, ohetransform], axis=1).drop(columns=['MaritalStatus','PerformanceRating','EnvironmentSatisfaction'])
X_test.head()


/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:975: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:975: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


HourlyRate  Age  YearsAtCompany  DistanceFromHome  YearsWithCurrManager  \
1291          58   37              10                10                     8   
1153          70   18               0                 3                     0   
720           48   30               5                22                     1   
763           87   34               1                10                     0   
976           68   56              19                23                     9   

      YearsSinceLastPromotion  PercentSalaryHike  MaritalStatus_Divorced  \
1291                        0                 15                     0.0   
1153                        0                 12                     0.0   
720                         0                 11                     0.0   
763                         0                 19                     0.0   
976                        15                 12                     0.0   

      MaritalStatus_Married  MaritalStatus_Single  PerformanceRating_3  \
1291                    0.0                   1.0                  1.0   
1153                    0.0                   1.0                  1.0   
720                     1.0                   0.0                  1.0   
763                     1.0                   0.0                  1.0   
976                     1.0                   0.0                  1.0   

      PerformanceRating_4  EnvironmentSatisfaction_1  \
1291                  0.0                        0.0   
1153                  0.0                        0.0   
720                   0.0                        1.0   
763                   0.0                        0.0   
976                   0.0                        0.0   

      EnvironmentSatisfaction_2  EnvironmentSatisfaction_3  \
1291                        0.0                        0.0   
1153                        1.0                        0.0   
720                         0.0                        0.0   
763                         0.0                        1.0   
976                         0.0                        0.0   

      EnvironmentSatisfaction_4  
1291                        1.0  
1153                        0.0  
720                         0.0  
763                         0.0  
976                         1.0

In [19]:
# Create a StandardScaler
scaler = StandardScaler()

# Fit the StandardScaler to the training data
X_scaler = scaler.fit(X_train)

# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)



In [20]:
# Create a OneHotEncoder for the Department column
ohe = OneHotEncoder(handle_unknown='ignore', sparse=False).set_output(transform="pandas")

# Fit the encoder to the training data
ohetransform = ohe.fit_transform(y_train[['Department']])
ohetransform.head()

# Create two new variables by applying the encoder
# to the training and testing data
y_train_department = pd.concat([y_train, ohetransform], axis=1).drop(columns=['Department','Attrition'])
y_train_department.head()


/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:975: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


Department_Human Resources  Department_Research & Development  \
464                          0.0                                1.0   
512                          0.0                                1.0   
480                          0.0                                0.0   
680                          0.0                                1.0   
1023                         0.0                                1.0   

      Department_Sales  
464                0.0  
512                0.0  
480                1.0  
680                0.0  
1023               0.0

In [21]:
# Create a OneHotEncoder for the Attrition column
ohe = OneHotEncoder(handle_unknown='ignore', sparse=False).set_output(transform="pandas")

# Fit the encoder to the training data
ohetransform = ohe.fit_transform(y_train[['Attrition']])
ohetransform.head()

# Create two new variables by applying the encoder
# to the training and testing data
y_train_attrition = pd.concat([y_train, ohetransform], axis=1).drop(columns=['Attrition', 'Department'])
y_train_attrition.head()


/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:975: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


Attrition_No  Attrition_Yes
464            1.0            0.0
512            1.0            0.0
480            0.0            1.0
680            1.0            0.0
1023           1.0            0.0

## Create, Compile, and Train the Model

In [22]:
# Find the number of columns in the X training data
columns = X_train_scaled.shape[1]

# Create the input layer
input_layer = Input(shape=(columns,), name='input_features')

# Create at least two shared layers
shared_layer1 = Dense(64, activation='relu')(input_layer)
shared_layer2 = Dense(32, activation='relu')(shared_layer1)

In [23]:
# Create a branch for Department
# with a hidden layer and an output layer

# Create the hidden layer
department_hidden = Dense(16, activation='relu')(shared_layer2)

# Create the output layer
department_output = Dense(3, activation='softmax', name='department_output')(department_hidden)


In [34]:
# Create a branch for Attrition
# with a hidden layer and an output layer

# Create the hidden layer
attrition_hidden = Dense(16, activation='relu')(shared_layer2)

# Create the output layer
attrition_output = Dense(2, activation='softmax', name='attrition_output')(attrition_hidden)



In [35]:
# Create the model
model = Model(inputs=input_layer, outputs=[department_output, attrition_output])

# Compile the model
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics={'department_output': 'accuracy',
                       'attrition_output': 'accuracy'})

# Summarize the model
model.summary()


Model: "functional_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)              ┃ Output Shape           ┃        Param # ┃ Connected to           ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━┩
│ input_features            │ (None, 16)             │              0 │ -                      │
│ (InputLayer)              │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense_3 (Dense)           │ (None, 64)             │          1,088 │ input_features[0][0]   │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense_4 (Dense)           │ (None, 32)             │          2,080 │ dense_3[0][0]          │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense_5 (Dense)           │ (None, 16)             │            528 │ dense_4[0][0]          │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense_9 (Dense)           │ (None, 16)             │            528 │ dense_4[0][0]          │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ department_output (Dense) │ (None, 3)              │             51 │ dense_5[0][0]          │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ attrition_output (Dense)  │ (None, 2)              │             34 │ dense_9[0][0]          │
└───────────────────────────┴────────────────────────┴────────────────┴────────────────────────┘

 Total params: 4,309 (16.83 KB)

 Trainable params: 4,309 (16.83 KB)

 Non-trainable params: 0 (0.00 B)

In [39]:
# Train the model
model.fit(X_train_scaled, [y_train_department, y_train_attrition], epochs=200, batch_size=32, validation_split=0.2, verbose=1)


Epoch 1/200
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - attrition_output_accuracy: 0.9885 - department_output_accuracy: 0.8905 - loss: 0.3586 - val_attrition_output_accuracy: 0.8100 - val_department_output_accuracy: 0.5520 - val_loss: 2.7833
Epoch 2/200
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - attrition_output_accuracy: 0.9784 - department_output_accuracy: 0.8740 - loss: 0.3830 - val_attrition_output_accuracy: 0.8054 - val_department_output_accuracy: 0.5430 - val_loss: 2.8619
Epoch 3/200
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - attrition_output_accuracy: 0.9838 - department_output_accuracy: 0.8952 - loss: 0.3477 - val_attrition_output_accuracy: 0.7964 - val_department_output_accuracy: 0.5566 - val_loss: 2.7840
Epoch 4/200
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - attrition_output_accuracy: 0.9908 - department_output_accuracy: 0.8698 - loss: 0.3870 - val_attrition_output_accuracy: 0.7964 - val_department_output_accuracy: 0.5249 - val_loss: 2.8401
Epoch 5/200
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 9

In [37]:
# Evaluate the model with the testing data
# Fit the encoder to the training data
ohetransform = ohe.fit_transform(y_test[['Department']])
# Create two new variables by applying the encoder
# to the training and testing data
y_test_department = pd.concat([y_test, ohetransform], axis=1).drop(columns=['Department','Attrition'])

# Fit the encoder to the training data
ohetransform = ohe.fit_transform(y_test[['Attrition']])
ohetransform.head()

# Create two new variables by applying the encoder
# to the training and testing data
y_test_attrition = pd.concat([y_test, ohetransform], axis=1).drop(columns=['Attrition', 'Department'])

model_loss, department_accuracy, attrition_accuracy = model.evaluate(X_test_scaled, [y_test_department, y_test_attrition])

12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - attrition_output_accuracy: 0.7514 - department_output_accuracy: 0.5374 - loss: 3.5391 


/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:975: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:975: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


In [38]:
# Print the accuracy for both department and attrition
print(f"Department Accuracy: {department_accuracy}")
print(f"Attrition Accuracy: {attrition_accuracy}")


Department Accuracy: 0.77173912525177
Attrition Accuracy: 0.5597826242446899


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1.
2.
3.

1. I don't believe Accuracy is the best metric because when you have on-hot encoded data, you have '1's and '0's - so it is all or nothing. for example if you have a predition as a whole number - say 35 feet - being 80% accurate will indicate that you are 30 or 40 feet. But a 70% of all or nothing does not tell you much or which items are accurate. You can surmise that the whole data is 55 % accurate (for the attrition) but it does not tell you which ones are accurate and which ones are not, so there is a limitation

2. I used 'soft max' as an activation function in the 2 output layers. Soft Max is appropriate when dealign with catergorical data that has been one-hot encoded list the 'Attrition' and 'Department' items.

3. Could use some of the methods we reviewed earlir to optimize the number of neurons and layers in the network.